src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -2,22 +2,134 @@
 
 import com.epam.aidial.core.ProxyContext;
 import com.epam.aidial.core.config.ApiKeyData;
+import com.epam.aidial.core.util.ProxyUtil;
 import com.google.common.annotations.VisibleForTesting;
 import io.vertx.core.Future;
 import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import lombok.extern.slf4j.Slf4j;
+import org.redisson.api.RTopic;
+import org.redisson.api.RedissonClient;
+import org.redisson.client.codec.StringCodec;
 
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.LongSupplier;
 
+@Slf4j
 public class TokenStatsTracker {
+
+
     private final Map<String, TraceContext> traceIdToContext = new ConcurrentHashMap<>();
```

Comment:
In case one of backends fails not publishing an end of span event, how do we clean this map?

Reply:
You mean a case when the Core is alive but publishing doesn't work for some reason, e.g. bug in the code or Redis is not functional. Right?

We don't clean the map in this case.

In my opinion that's a bug in Core or infrastructure issue. Those traces will stay forever.

If you believe that's a strong case we can introduce a job that scans the map to find obsolete traces and remove them after some timeout, e.g. 1 min.